### PR TITLE
Router

### PR DESCRIPTION
**Pull request by quirk0o**
@adamzima @s1mplex @pbzymek @mhib @lkudyba @jtkpiotr

### CODE REVIEW

File: client/app/components/App.jsx [hunk 1]
```diff
@@ -42,6 +47,13 @@ export default class AppContainer extends React.Component {
     }
   }
 
+  componentDidUpdate(prevProps) {
+    const currentQuery = this.props.location.query;
+    const prevQuery = prevProps.location.query;
+    if (prevQuery && currentQuery && prevQuery.date === currentQuery.date) return;
```
Comment (adamzima) [inline client/app/components/App.jsx:56]: Do we care only for date param right now?

File: client/app/components/ReactRoot.jsx [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import { Router, Route, browserHistory } from 'react-router';
+
+import App from './App';
+
+const wrapComponent = (Component, ownProps) => (props) => (
```
Comment (adamzima) [inline client/app/components/ReactRoot.jsx:6]: :+1:

File: client/app/components/App.jsx [hunk 3]
```diff
@@ -1,24 +1,29 @@
...
-export default class AppContainer extends React.Component {
+export default class App extends React.Component {
   static propTypes = {
-    initialEvents:       React.PropTypes.array,
-    date:                React.PropTypes.string,
-    notificationTimeout: React.PropTypes.number
+    initialEvents:       PropTypes.array,
+    weekLength:          PropTypes.number,
+    location:            PropTypes.shape({
+      query: PropTypes.shape({
```
Comment (adamzima) [inline client/app/components/App.jsx]: brrr me not like this indendation

File: client/app/components/layout/SideNav.jsx [hunk 4]
```diff
@@ -25,27 +27,26 @@ export default class SideNav extends React.Component {
                 onClick={this._handleCreateEvent}>
           Create Event
         </Button>
-        <Button href={"/"}
-                className="btn-block">Home</Button>
-        <Button href={this._dateParam(this._nextWeek())}
-                className="btn-block">Next Week</Button>
-        <Button href={this._dateParam(this._previousWeek())}
-                className="btn-block">Previous Week</Button>
+        <IndexLinkContainer to="/" active={false}>
```
Comment (adamzima) [inline client/app/components/layout/SideNav.jsx:30]: Do we need to pass active here?

File: client/app/components/layout/SideNav.jsx [hunk 5]
```diff
@@ -25,27 +27,26 @@ export default class SideNav extends React.Component {
...
+    return moment(this.props.date).add(1, 'weeks').format(DATE_PARAM_FORMAT);
   }
 
   _previousWeek() {
-    return moment(this.props.date).subtract(1, 'weeks');
-  }
-
-  _dateParam(date) {
-    return `/?date=${date.format('YYYY-MM-DD')}`;
+    return moment(this.props.date).subtract(1, 'weeks').format(DATE_PARAM_FORMAT);
```
Comment (adamzima) [inline client/app/components/layout/SideNav.jsx]:
``` javascript
_previousWeek() {
  _formattedDate(moment(this.props.date).subtract(1, 'weeks'))
}

_formattedDate(date) {
  date.format(DATE_PARAM_FORMAT)
}
```

Comment (adamzima) [inline client/app/components/layout/SideNav.jsx]: Or event better move this function to `DateHelper`. What's your opinion?

File: client/package.json [hunk 6]
```diff
@@ -7,7 +7,7 @@
   },
   "scripts": {
     "build:client": "NODE_ENV=production webpack --config webpack.build.config.js",
-    "test": "NODE_ENV=test mocha \"test/**/*.jsx\"",
+    "test": "NODE_ENV=test mocha \"test/**/*.js?(x)\"",
```
Comment (adamzima) [inline client/package.json:10]: :)

File: client/app/components/App.jsx (hunk 1 above)
Comment (quirk0o) [inline client/app/components/App.jsx:56]: yes ^^

File: client/app/components/layout/SideNav.jsx [hunk 7]
```diff
@@ -1,12 +1,14 @@
...
 import { Button } from 'react-bootstrap';
+import { LinkContainer, IndexLinkContainer } from 'react-router-bootstrap';
+import { DATE_PARAM_FORMAT } from 'helpers/DateHelper';
 
 import RefreshButton from './RefreshButton';
 
 export default class SideNav extends React.Component {
   static propTypes = {
-    date:      React.PropTypes.string.isRequired,
+    date:      PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.string]).isRequired,
```
Comment (madejejej) [inline client/app/components/layout/SideNav.jsx]: Can we just inject one of these types, eg. Date? It's always confusing when you have to handle more than one type

File: client/app/components/calendar/CalendarHeader.jsx [hunk 8]
```diff
@@ -8,7 +8,7 @@ const CalendarHeader = (props) => (
 );
 
 CalendarHeader.propTypes = {
-  day:        React.PropTypes.string.isRequired,
+  day:        React.PropTypes.oneOfType([React.PropTypes.instanceOf(Date), React.PropTypes.string]).isRequired,
```
Comment (madejejej) [inline client/app/components/calendar/CalendarHeader.jsx]: Can we just inject one of these types, eg. Date? It's always confusing when you have to handle more than one type

File: client/app/components/layout/SideNav.jsx (hunk 7 above)
Comment (adamzima) [inline client/app/components/layout/SideNav.jsx]: :+1:

File: client/app/components/layout/SideNav.jsx (hunk 4 above)
Comment (quirk0o) [inline client/app/components/layout/SideNav.jsx:30]: yes ^^

File: client/.eslintrc.yml [hunk 9]
```diff
@@ -36,6 +36,7 @@ rules:
   max-len: [2, 120]
   no-param-reassign: 0
   one-var: 0
+  no-eq-null: 0
```
Comment (adamzima) [inline client/.eslintrc.yml]: why :D?

Comment (adamzima) [inline client/.eslintrc.yml]: nope

File: client/app/components/App.jsx [hunk 10]
```diff
@@ -83,16 +106,30 @@ export default class AppContainer extends React.Component {
     );
   }
 
+  _dateOrNow() {
+    const query = this.props.location.query;
+    if (query && query.date) {
+      return moment(query.date);
+    }
+    return moment();
```
Comment (adamzima) [inline client/app/components/App.jsx]: How about passing beginning of day.

**Pull request merged**